## Importing Libraires

In [8]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Loading dataset

In [9]:
df = pd.read_csv('student-por.csv')

# 1. Dataset Understanding

### How many rows and columns are present in the dataset?

In [10]:
df.shape

(649, 33)

"There are 649 rows are 33 columns in the dataset"

### What are the names and meanings of all attributes?

In [11]:
df.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='str')

| # | Attribute | Meaning |
|---|---|---|
| 1 | `school` | Student's school — GP (Gabriel Pereira) or MS (Mousinho da Silveira) |
| 2 | `sex` | Student's gender — M (Male) or F (Female) |
| 3 | `age` | Student's age (15 to 22) |
| 4 | `address` | Home location — U (Urban) or R (Rural) |
| 5 | `famsize` | Family size — LE3 (3 or less) or GT3 (more than 3) |
| 6 | `Pstatus` | Parents living together — T (Together) or A (Apart) |
| 7 | `Medu` | Mother's education level (0=none, 1=primary, 2=middle, 3=secondary, 4=higher) |
| 8 | `Fedu` | Father's education level (same scale) |
| 9 | `Mjob` | Mother's job (teacher, health, services, at_home, other) |
| 10 | `Fjob` | Father's job (same categories) |
| 11 | `reason` | Reason for choosing this school (home, reputation, course, other) |
| 12 | `guardian` | Student's guardian (mother, father, other) |
| 13 | `traveltime` | Home to school travel time (1=<15min, 2=15-30min, 3=30-60min, 4=>1hr) |
| 14 | `studytime` | Weekly study time (1=<2hrs, 2=2-5hrs, 3=5-10hrs, 4=>10hrs) |
| 15 | `failures` | Number of past class failures (0 to 3) |
| 16 | `schoolsup` | Extra educational support from school (yes/no) |
| 17 | `famsup` | Educational support from family (yes/no) |
| 18 | `paid` | Extra paid classes (yes/no) |
| 19 | `activities` | Extracurricular activities (yes/no) |
| 20 | `nursery` | Attended nursery school (yes/no) |
| 21 | `higher` | Wants to pursue higher education (yes/no) |
| 22 | `internet` | Internet access at home (yes/no) |
| 23 | `romantic` | In a romantic relationship (yes/no) |
| 24 | `famrel` | Quality of family relationships (1=very bad, 5=excellent) |
| 25 | `freetime` | Free time after school (1=very low, 5=very high) |
| 26 | `goout` | Going out with friends (1=very low, 5=very high) |
| 27 | `Dalc` | Workday alcohol consumption (1=very low, 5=very high) |
| 28 | `Walc` | Weekend alcohol consumption (1=very low, 5=very high) |
| 29 | `health` | Current health status (1=very bad, 5=very good) |
| 30 | `absences` | Number of school absences (0 to 93) |
| 31 | `G1` | First term grade (0–20) |
| 32 | `G2` | Second term grade (0–20) |
| 33 | `G3` | Final grade (0–20) — **Target Variable** |

### What are the data types of each column?

In [16]:
print("Data Types of Each Column:")
print(df.dtypes)
print(f"\nTotal Numerical Columns : {df.select_dtypes(include='int64').shape[1]}")
print(f"Total Categorical Columns: {df.select_dtypes(include='str').shape[1]}")

Data Types of Each Column:
school          str
sex             str
age           int64
address         str
famsize         str
Pstatus         str
Medu          int64
Fedu          int64
Mjob            str
Fjob            str
reason          str
guardian        str
traveltime    int64
studytime     int64
failures      int64
schoolsup       str
famsup          str
paid            str
activities      str
nursery         str
higher          str
internet        str
romantic        str
famrel        int64
freetime      int64
goout         int64
Dalc          int64
Walc          int64
health        int64
absences      int64
G1            int64
G2            int64
G3            int64
dtype: object

Total Numerical Columns : 16
Total Categorical Columns: 17


### Are there any missing or duplicate values in the dataset?

In [17]:
print("Missing Values in Each Column:")
print(df.isnull().sum())
print(f"\nTotal Missing Values: {df.isnull().sum().sum()}")
print(f"Duplicate Rows: {df.duplicated().sum()}")

Missing Values in Each Column:
school        0
sex           0
age           0
address       0
famsize       0
Pstatus       0
Medu          0
Fedu          0
Mjob          0
Fjob          0
reason        0
guardian      0
traveltime    0
studytime     0
failures      0
schoolsup     0
famsup        0
paid          0
activities    0
nursery       0
higher        0
internet      0
romantic      0
famrel        0
freetime      0
goout         0
Dalc          0
Walc          0
health        0
absences      0
G1            0
G2            0
G3            0
dtype: int64

Total Missing Values: 0
Duplicate Rows: 0


## Missing & Duplicate Values

### Missing Values
- Checked using `df.isnull().sum()`
- **Result: 0 missing values** in all 33 columns
- Dataset is complete — no imputation needed

### Duplicate Rows
- Checked using `df.duplicated().sum()`
- **Result: 0 duplicate rows** found
- Every row represents a unique student

### Conclusion
> The dataset is clean with no missing or duplicate values.
> We can proceed directly to preprocessing.

### Which variables are numerical and which are categorical?

In [18]:
numerical = df.select_dtypes(include='int64').columns.tolist()
categorical = df.select_dtypes(include='str').columns.tolist()

print(f"Numerical Variables ({len(numerical)}):")
print(numerical)

print(f"\nCategorical Variables ({len(categorical)}):")
print(categorical)

Numerical Variables (16):
['age', 'Medu', 'Fedu', 'traveltime', 'studytime', 'failures', 'famrel', 'freetime', 'goout', 'Dalc', 'Walc', 'health', 'absences', 'G1', 'G2', 'G3']

Categorical Variables (17):
['school', 'sex', 'address', 'famsize', 'Pstatus', 'Mjob', 'Fjob', 'reason', 'guardian', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery', 'higher', 'internet', 'romantic']


### Summary

| Type | Count | Action |
|---|---|---|
| Numerical | 16 | Ready to use directly in model |
| Categorical | 17 | Will be encoded using Label Encoding in preprocessing |

# 2 Data Preprocessing

### How will missing values (if any) be handled?

- Checked in Dataset Understanding section using `df.isnull().sum()`
- Result: No missing values found in any column
- Therefore, no imputation or dropping of rows was required

### How will categorical variables be converted into numerical form?

In [23]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()
cat_cols = df.select_dtypes(include='str').columns

for col in cat_cols:
    df[col] = le.fit_transform(df[col])

print("Encoding Complete ✅")
df.head()

Encoding Complete ✅


,school,sex,age,address,famsize,Pstatus,Medu,Fedu,Mjob,Fjob,...,famrel,freetime,goout,Dalc,Walc,health,absences,G1,G2,G3
0,0,0,18,1,0,0,4,4,0,4,...,4,3,4,1,1,3,4,0,11,11
1,0,0,17,1,0,1,1,1,0,2,...,5,3,3,1,1,3,2,9,11,11
2,0,0,15,1,1,1,1,1,0,2,...,4,3,2,2,3,3,6,12,13,12
3,0,0,15,1,0,1,4,2,1,3,...,3,2,2,1,1,5,0,14,14,14
4,0,0,16,1,0,1,3,3,2,2,...,4,3,2,1,2,5,0,11,13,13


### Encoded Columns & Their Values

| Column | Original Values | Encoded Values |
|---|---|---|
| `school` | GP, MS | 0, 1 |
| `sex` | F, M | 0, 1 |
| `address` | R, U | 0, 1 |
| `famsize` | GT3, LE3 | 0, 1 |
| `Pstatus` | A, T | 0, 1 |
| `Mjob` | at_home, health, other, services, teacher | 0, 1, 2, 3, 4 |
| `Fjob` | at_home, health, other, services, teacher | 0, 1, 2, 3, 4 |
| `reason` | course, home, other, reputation | 0, 1, 2, 3 |
| `guardian` | father, mother, other | 0, 1, 2 |
| `schoolsup` | no, yes | 0, 1 |
| `famsup` | no, yes | 0, 1 |
| `paid` | no, yes | 0, 1 |
| `activities` | no, yes | 0, 1 |
| `nursery` | no, yes | 0, 1 |
| `higher` | no, yes | 0, 1 |
| `internet` | no, yes | 0, 1 |
| `romantic` | no, yes | 0, 1 |